gitlab/resource_gitlab_repository_file.go

```diff
@@ -0,0 +1,219 @@
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"author_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"content": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateBase64Content,
+			},
+			"commit_message": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"overwrite_on_create": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceGitlabRepositoryFileCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	filePath := d.Get("file_path").(string)
+
+	var existingRepositoryFile *gitlab.File
+	if d.Get("overwrite_on_create").(bool) {
+		readOptions := &gitlab.GetFileOptions{
+			Ref: gitlab.String(d.Get("branch").(string)),
+		}
+
+		existingRepositoryFile, _, _ = client.RepositoryFiles.GetFile(project, filePath, readOptions)
+	}
+
+	var filePathForId string
+	if existingRepositoryFile == nil {
+		options := &gitlab.CreateFileOptions{
+			Branch:        gitlab.String(d.Get("branch").(string)),
+			Encoding:      gitlab.String(encoding),
+			AuthorEmail:   gitlab.String(d.Get("author_email").(string)),
+			AuthorName:    gitlab.String(d.Get("author_name").(string)),
+			Content:       gitlab.String(d.Get("content").(string)),
+			CommitMessage: gitlab.String(d.Get("commit_message").(string)),
+		}
+		if startBranch, ok := d.GetOk("start_branch"); ok {
+			options.StartBranch = gitlab.String(startBranch.(string))
+		}
+
+		repositoryFile, _, err := client.RepositoryFiles.CreateFile(project, filePath, options)
+		if err != nil {
+			return err
+		}
+		filePathForId = repositoryFile.FilePath
+	} else {
+		options := &gitlab.UpdateFileOptions{
+			Branch:        gitlab.String(d.Get("branch").(string)),
+			Encoding:      gitlab.String(encoding),
+			AuthorEmail:   gitlab.String(d.Get("author_email").(string)),
+			AuthorName:    gitlab.String(d.Get("author_name").(string)),
+			Content:       gitlab.String(d.Get("content").(string)),
+			CommitMessage: gitlab.String(d.Get("commit_message").(string)),
+			LastCommitID:  gitlab.String(existingRepositoryFile.LastCommitID),
+		}
+		if startBranch, ok := d.GetOk("start_branch"); ok {
+			options.StartBranch = gitlab.String(startBranch.(string))
+		}
+
+		repositoryFile, _, err := client.RepositoryFiles.UpdateFile(project, filePath, options)
+		if err != nil {
+			return err
+		}
+		filePathForId = repositoryFile.FilePath
+	}
+
+	d.SetId(filePathForId)
+	return resourceGitlabRepositoryFileRead(d, meta)
+}
+
+func resourceGitlabRepositoryFileRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	filePath := d.Id()
+	options := &gitlab.GetFileOptions{
+		Ref: gitlab.String(d.Get("branch").(string)),
+	}
+
+	repositoryFile, _, err := client.RepositoryFiles.GetFile(project, filePath, options)
+	if err != nil {
```

Comment:
it would be great if we could handle the 404 error case a little differently, as in the following example here from the AWS provider:
https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/resource_aws_ami.go#L344

It uses `d.SetId("")` to indicate that the resource no longer exists instead of throwing the API error back to terraform.

From my perspective, this behaviour is highly desirable because it prevents the terraform drift detection stage from crashing if someone manually removed the file before, or the underlying project doesn't exist anymore.

This behavior is present in the entire AWS provider, but missing in the entire Gitlab provider 🙂 not sure what's the best approach here... Start setting an example in this resource or doing a complete refactoring at once...

Reply:
I've created an issue from this for discussion here: #726 

@mkjmdski what do you think? Should I implement something like this in this PR already or does it make sense to refactor the entire provider all at once?

Reply:
I've encountered this issue few times but didn't know the root cause. I think that making this refactor just for this specific resource in this pull request is good idea, and adding it to the others as a part of separated issue is another good idea too :D